src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,157 @@
+    if ((msgIndex < 0) || (msgIndex >= m_messages.size()))
+        return {};
+
+    const Message &message = m_messages[msgIndex];
+    switch (role) {
+    case Qt::ForegroundRole:
+        return message.foregroundRole;
+    case Qt::UserRole:
+        return message.userRole;
+    case TimeRole:
+        return message.time;
+    case MessageRole:
+        return message.message;
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(const BaseLogModel::Message& message)
+{
+    // if row is inserted on filled up buffer, the size will not change but because of calling of beginInsertRows function
+    // we'll have ghost rows.
+    if (m_messages.size() == MAX_LOG_MESSAGES) {
+        const int lastMessage = m_messages.size() - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        m_messages.pop_front();
+        endRemoveRows();
+    }
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    m_messages.push_back(message);
+    endInsertRows();
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_messages.clear();
+    endResetModel();
+}
+
+LogMessageModel::LogMessageModel(QObject *parent)
+    : BaseLogModel(parent)
+{
+    for (const Log::Msg &msg : asConst(Logger::instance()->getMessages()))
+        handleNewMessage(msg);
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage, Qt::QueuedConnection);
+}
+
+void LogMessageModel::handleNewMessage(const Log::Msg &log)
+{
+    Message message;
+
+    message.time = QDateTime::fromMSecsSinceEpoch(log.timestamp).toString(Qt::SystemLocaleShortDate);
+    message.message = log.message;
+
+    switch (log.type) {
+    // The RGB QColor constructor is used for performance
+    case Log::NORMAL:
+        message.foregroundRole = QApplication::palette().color(QPalette::WindowText);
+        break;
+    case Log::INFO:
+        message.foregroundRole = QColor(0, 0, 255); // blue
+        break;
+    case Log::WARNING:
+        message.foregroundRole = QColor(255, 165, 0);  // orange
+        break;
+    case Log::CRITICAL:
+        message.foregroundRole = QColor(255, 0, 0);  // red
+        break;
+    default:
+        Q_ASSERT(false);
+        break;
+    }
+
+    message.userRole = log.type;
+    addNewMessage(message);
+}
+
+LogPeerModel::LogPeerModel(QObject *parent)
+    : BaseLogModel(parent)
+{
+    for (const Log::Peer &peer : asConst(Logger::instance()->getPeers()))
+        handleNewMessage(peer);
+    connect(Logger::instance(), &Logger::newLogPeer, this, &LogPeerModel::handleNewMessage, Qt::QueuedConnection);
+}
+
+void LogPeerModel::handleNewMessage(const Log::Peer &peer)
+{
+    const QString time = QDateTime::fromMSecsSinceEpoch(peer.timestamp).toString(Qt::SystemLocaleShortDate);
+    const QString message = (peer.blocked
+                             ? tr("%1 was blocked %2", "x.y.z.w was blocked").arg(peer.ip, peer.reason)
```

Comment:
Format string looks ugly.

Reply:
I changed it to what it was originally in `ExecutionLogWidget`, but don't know what to do further.